JIT: inline shared generics with runtime lookups inside

EgorBo opened the pull request:
Closes https://github.com/dotnet/runtime/issues/81432
Closes https://github.com/dotnet/runtime/issues/8662
Closes https://github.com/dotnet/runtime/issues/65815
_(probably more issues, need to check)_

Based on https://github.com/dotnet/runtime/pull/81635.

This PR implements @jkotas's [idea](https://github.com/dotnet/runtime/issues/81432#issuecomment-1411268454) to inline shared generics which require runtime lookups.
Example:

```cs
[MethodImpl(MethodImplOptions.NoInlining)]
bool Test<T>() => Callee<T>();

bool Callee<T>() => typeof(T).IsValueType;
```
Codegen for `Test<string>` in Main:
```asm
; Assembly listing for method Program:Test[System.__Canon]():bool:this
       push     rbx
       sub      rsp, 48
       mov      qword ptr [rsp+0x28], rdx
       mov      rbx, rcx
       mov      rcx, qword ptr [rdx+0x10]
       mov      rax, qword ptr [rcx+0x10]
       test     rax, rax
       je       SHORT G_M000_IG04
       mov      rdx, rax
       jmp      SHORT G_M000_IG05
G_M000_IG04:
       mov      rcx, rdx
       mov      rdx, 0x7FFCC5FE7088
       call     CORINFO_HELP_RUNTIMEHANDLE_METHOD
       mov      rdx, rax
G_M000_IG05:
       mov      rcx, rbx
       call     [Program:Callee[System.__Canon]():bool:this]   ;;;  <---- not inlined
       nop      
       add      rsp, 48
       pop      rbx
       ret      
; Total bytes of code 68
```
Codegen for `Test<string>` in this PR:
```asm
; Assembly listing for method Program:Test[System.__Canon]():ubyte:this
       xor      eax, eax
       ret
```

I think the JIT side is ok (I just need to double check various "this context load is invariant" places), but I am not sure I implemented the VM side (only CoreCLR for now) correctly.

jkotas commented on src/coreclr/inc/corinfo.h:
```diff
@@ -1227,6 +1227,7 @@ enum CORINFO_RUNTIME_LOOKUP_KIND
 struct CORINFO_LOOKUP_KIND
 {
     bool                        needsRuntimeLookup;
+    bool                        inlinedLookup;
```
I do not understand why this new flag is needed. What is not going to work without it?

EgorBo replied:
@jkotas my initial version didn't need it at all, I added it for additional asserts in JIT, I think my final version won't have it

EgorBo committed:
Simpler change

EgorBo commented:
/azp run runtime-coreclr outerloop, runtime-coreclr gcstress0x3-gcstress0xc

EgorBo commented:
@MihuBot

jakobbotsch commented on src/coreclr/jit/importercalls.cpp:
```diff
@@ -6730,6 +6730,17 @@ void Compiler::impMarkInlineCandidate(GenTree*               callNode,
     // candidate, we're done.
     if (call->IsInlineCandidate() || !call->IsGuardedDevirtualizationCandidate())
     {
+        // Runtime lookup is expected to be spilled into a temp for inline candidates.
+        CallArg* instParam = call->gtArgs.FindWellKnownArg(WellKnownArg::InstParam);
+        if ((instParam != nullptr) && instParam->GetNode()->OperIs(GT_RUNTIMELOOKUP))
+        {
+            const unsigned lookupTmp = lvaGrabTemp(true DEBUGARG("spilling runtimelookup"));
+            impStoreTemp(lookupTmp, instParam->GetNode(), CHECK_SPILL_NONE);
+            GenTreeLclVar* lcl = gtNewLclvNode(lookupTmp, TYP_I_IMPL);
+            lcl->gtFlags |= GTF_VAR_CONTEXT;
+            instParam->SetEarlyNode(lcl);
+        }
```
Is it ok to reorder its evaluation with other arguments?

jakobbotsch replied:
I mean both the IL stack but also the other operands of the call. Those will get reordered with the runtime lookup.

The reordering might still be illegal even if the runtime lookup is invariant as a whole, for example if runtime lookups can throw exceptions (which presumably they can if they result in type loading?).

EgorBo replied:
I'll check how diffs look like with proper spilling

jakobbotsch replied:
Outside x86 the instantiation parameter is inserted after "this" but before the rest of the arguments, so if the runtime lookup can throw exceptions it seems like we're already getting it wrong today.

EgorBo replied:
I think it's similar to how we handle OOM side-effects from allocations

jkotas replied:
Yes, the runtime lookup can throw exceptions for invalid programs or for out-of-memory situations. We do not guarantee the exact position where the exception is going to be thrown.

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -1570,6 +1570,19 @@ GenTree* Compiler::getRuntimeContextTree(CORINFO_RUNTIME_LOOKUP_KIND kind)
     {
         assert(kind == CORINFO_LOOKUP_METHODPARAM || kind == CORINFO_LOOKUP_CLASSPARAM);
 
+        if (compIsForInlining() && (kind == CORINFO_LOOKUP_METHODPARAM))
```
Why is the `kind == CORINFO_LOOKUP_METHODPARAM` check needed? Does it mean we won't be able to inline non-generic static methods in generic classes?

EgorBo replied:
I was think of this check as a TP check to avoid going into FindWellKnownArg. I don't see why would it prevent from inlining a non-generic static methods in generic classes?

jakobbotsch replied:
Then:
```suggestion
        if (compIsForInlining() && ((kind == CORINFO_LOOKUP_CLASSPARAM) || (kind == CORINFO_LOOKUP_METHODPARAM)))
```

I think the instantiation parameter for a non-generic static method in a generic class is going to have `kind == CORINFO_LOOKUP_CLASSPARAM`, but I could be wrong.

There's still potential quadratic complexity for pathological cases here. Perhaps it would be a bit more elegant to have some `InlineInfo::instArgLclNum` directly? We already recognize it here, so it would just require saving the local:
https://github.com/dotnet/runtime/blob/7c0ed473e76a1861568fa2269f855e2d2e3f16cc/src/coreclr/jit/importer.cpp#L12774-L12787

EgorBo committed:
Address feedback

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -3086,24 +3086,28 @@ void CEEInfo::ComputeRuntimeLookupForSharedGenericToken(DictionaryEntryKind entr
+    MethodDesc* pContextMD = GetMethodFromContext(pResolvedToken->tokenContext);
+    if (pContextMD == nullptr)
     {
+        // Class context is not yet supported
```
What's missing for this?

EgorBo replied:
Working on it, should be this VM function specific change.
+ need to unmark a couple of trees as invariant in jit

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -1548,31 +1548,39 @@ GenTree* Compiler::impMethodPointer(CORINFO_RESOLVED_TOKEN* pResolvedToken, CORI
+        {
+            // Grab the generic context from the callsite for current inlinee.
+            // For runtime lookups, the generic context is always spilled to a local
+            assert(impInlineInfo->iciCallInstParam->OperIs(GT_LCL_VAR));
```
IMO, it would be cleaner to do this assert once and store the local number instead of the `GenTree*`.

EgorBo replied:
I don't think I can assert untill I need an actual runtime lookup. the context may still be GT_CNS_INT/IND for other cases. the current shape allows me better validation

jakobbotsch commented on src/coreclr/jit/inline.h:
```diff
@@ -702,9 +702,10 @@ struct InlineInfo
-    Statement*   iciStmt;  // The statement iciCall is in.
-    BasicBlock*  iciBlock; // The basic block iciStmt is in.
+    GenTreeCall* iciCall;          // The GT_CALL node to be inlined.
+    GenTree*     iciCallInstParam; // Cached InstParam arg (if exists) in iciCall
```
Super nit: Put this next to the fields that relate to the arguments further above?

jakobbotsch commented on src/coreclr/jit/importercalls.cpp:
```diff
@@ -6730,6 +6730,21 @@ void Compiler::impMarkInlineCandidate(GenTree*               callNode,
     {
+        // Runtime lookup is expected to be spilled into a temp for inline candidates.
+        CallArg* instParam = call->gtArgs.FindWellKnownArg(WellKnownArg::InstParam);
+        if ((instParam != nullptr) && instParam->GetNode()->OperIs(GT_RUNTIMELOOKUP))
```
The `OperIs(GT_RUNTIMELOOKUP)` check is odd to me. Can the check be `!OperIs(GT_LCL_VAR)` instead, to make the invariants more clear?

EgorBo replied:
Why would we want to spill GT_CNS_INT context? or any other non-RUNTIMELOOKUP one

jakobbotsch replied:
Ok, I think I misunderstood. So we will only call `getRuntimeContextTree` in the inlinee if we in the caller needed a runtime lookup for the instantiation arg?

It seems it would be overall better to make the handling uniform with how other inline arguments are handled to avoid the complicated contract here. `impInlineFetchArg` has a lot of smart handling in it, so maybe that would help with CQ as well (e.g. avoid the spilling in some common cases, like if the generic context ends up unused in the inlinee). The spilling would be handled only when necessary by `fgInlinePrependStatements`.

jakobbotsch replied:
Basically the idea would be:
1. Create a fully fledged `InlArgInfo` for the context argument. Store it separately from `inlArgInfo` in `InlineInfo`.
2. Refactor `impInlineFetchArg` to avoid needing `lclNum` (just requires passing the `InlArgInfo` and `InlLclVarInfo` entries directly)
3. Switch `getRuntimeContextTree` to use `impInlineFetchArg` on the `InlArgInfo` created for the context argument
4. Add handling for the new `InlArgInfo` to `fgInlinePrependStatements`, to make sure it's spilled if used or completely removed if unused.

Could be done in a separate PR...

EgorBo replied:
@jakobbotsch good idea! Addressed

EgorBo committed:
clean up

EgorBo committed:
Address feedback

EgorBo committed:
Clean up

EgorBo committed:
fix build

EgorBo committed:
clean up

EgorBo committed:
Class context?

EgorBo commented:
@jkotas one thing I struggle with currently is when I have `tokenContext` being class (CORINFO_CONTEXTFLAGS_CLASS) -- how do I know whether to use `CORINFO_LOOKUP_CLASSPARAM` or `CORINFO_LOOKUP_THISOBJ` ?

jkotas commented:
> when I have tokenContext being class (CORINFO_CONTEXTFLAGS_CLASS) -- how do I know whether to use CORINFO_LOOKUP_CLASSPARAM or CORINFO_LOOKUP_THISOBJ

It is not possible to tell from `CORINFO_CONTEXTFLAGS_CLASS` context alone. You need to have the current method to figure it out.

EgorBo commented:
> > when I have tokenContext being class (CORINFO_CONTEXTFLAGS_CLASS) -- how do I know whether to use CORINFO_LOOKUP_CLASSPARAM or CORINFO_LOOKUP_THISOBJ
> 
> It is not possible to tell from `CORINFO_CONTEXTFLAGS_CLASS` context alone. You need to have the current method to figure it out.

Current method as in the one we currently inline, right? not the `m_methodBeingCompiled` (which is the root method). So I presume I need to extend the JIT-EE interface

EgorBo commented:
Although, I am not sure I understand where should I pass it, ComputeRuntimeLookupForSharedGenericToken is not a direct jit-ee api..

EgorBo committed:
Disable class context

EgorBo committed:
Add containingMd to embedGenericHandle

jkotas commented:
Right, the problem is that `CORINFO_RESOLVED_TOKEN::tokenContext` does not carry information about the method. For non-generic methods in generic types, we use the generic type as the context. We are doing that because we do not always have a `MethodDesc*` that represents the method on the generic type to use as the context. It may be possible to make these method desc, but it would be wasteful.

`CORINFO_CONTEXT_HANDLE` represents method or type today. I think you need to make `CORINFO_CONTEXT_HANDLE` to represent a method and type tuple, at least on the paths that lead to `ComputeRuntimeLookupForSharedGenericToken`. `ComputeRuntimeLookupForSharedGenericToken` would then use the method to tell the runtimeLookupKind to use [here](https://github.com/dotnet/runtime/blob/6ebc8bd86dbc780b2a2a7daf3ab6020f9104f09e/src/coreclr/vm/jitinterface.cpp#L3096).

Changing `CORINFO_CONTEXT_HANDLE` into a struct that has two fields is probably the most straightforward way to do this.

EgorBo committed:
clean up

EgorBo committed:
initial NAOT support

EgorBo committed:
Implement NAOT

EgorBo commented:
> Changing `CORINFO_CONTEXT_HANDLE` into a struct that has two fields is probably the most straightforward way to do this.

I presume this adds 8 byte to a popular type that is used everywhere, I decided to propagate the containing Mth separately - it turned out to be easy, we need to compute the lookup from 3 places:
1) `getCallInfo` - it already had a "caller handle" parameter that represents exactly what I need - no change here
2) `embedGenericHandle` - I added "containingFtn" param here
3) `expandRawHandleIntrinsic` (NAOT-only API) - I added "containingFtn" param here as well

Now it seems to be passing CI and inlines way more shared generics (basically, all of them, we no longer give up on lookups)

An example (as I recall @stephentoub hit it somewhere relatively recently):
```cs
[MethodImpl(MethodImplOptions.NoInlining)]
int Foo<T>(Span<T> span) => Bar((ReadOnlySpan<T>)span); // forward to ROS<T> impl

int Bar<T>(ReadOnlySpan<T> span) => span.Length;
```
Codegen for `Foo<string>` in Main:
```asm
; Assembly listing for method Tests:Foo[System.__Canon](System.Span`1[System.__Canon])
       push     rbx
       sub      rsp, 64
       xor      eax, eax
       mov      qword ptr [rsp+0x28], rax
       mov      qword ptr [rsp+0x38], rdx
       mov      rbx, r8
       mov      rcx, qword ptr [rdx+0x38]
       mov      rax, qword ptr [rcx+0x10]
       test     rax, rax
       je       SHORT G_M21633_IG04
       mov      rdx, rax
       jmp      SHORT G_M21633_IG05
G_M21633_IG04:
       mov      rcx, rdx
       mov      rdx, 0x7FFC3F2C3498
       call     CORINFO_HELP_RUNTIMEHANDLE_METHOD
       mov      rdx, rax
G_M21633_IG05:
       lea      rcx, [rsp+0x28]
       mov      r8, rbx
       ;;;; NOT INLINED (requires runtime lookup)
       call     [System.Span`1[System.__Canon]:op_Implicit(System.Span`1[System.__Canon]):System.ReadOnlySpan`1[System.__Canon]]
       mov      eax, dword ptr [rsp+0x30]
       add      rsp, 64
       pop      rbx
       ret      
; Total bytes of code 83
```
New codegen:
```asm
; Assembly listing for method Tests:Foo[System.__Canon](System.Span`1[System.__Canon])
       mov      eax, dword ptr [r8+0x08]
       ret      
```

EgorBo committed:
Clean up

EgorBo committed:
Cleanup getRuntimeContextTree

EgorBo committed:
Merge branch 'main' into shared-generics-2

EgorBo committed:
Enable for R2R as well

EgorBo commented:
@MihuBot

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into shared-generics-2

jkotas commented on src/coreclr/inc/corinfo.h:
```diff
@@ -2221,6 +2221,7 @@ class ICorStaticInfo
     // should be looked up at runtime.
     virtual void expandRawHandleIntrinsic(
         CORINFO_RESOLVED_TOKEN *        pResolvedToken,
+        CORINFO_METHOD_HANDLE           containingFtn,
```
Should this be called `callerHandle`? It is what we call it in several existing methods.

EgorBo replied:
Addressed

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -3086,19 +3089,14 @@ void CEEInfo::ComputeRuntimeLookupForSharedGenericToken(DictionaryEntryKind entr
-        return;
-    }
+    bool inlinedLookup = pResolvedToken->tokenContext != METHOD_BEING_COMPILED_CONTEXT();
+    MethodDesc* pContextMD = inlinedLookup ? pContainingMD : GetMethodFromContext(pResolvedToken->tokenContext);
```
Use `pContainingMD` unconditionally? This should not need to special-case inlinedLookup vs. non-inlinedLookup.

Also, I think you can get rid of `METHOD_BEING_COMPILED_CONTEXT`. It should not be needed, and it is unnecessary complexity with your change.

EgorBo replied:
I think it should work, addressed

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -3086,19 +3089,14 @@ void CEEInfo::ComputeRuntimeLookupForSharedGenericToken(DictionaryEntryKind entr
     MethodTable* pContextMT = pContextMD->GetMethodTable();
-    bool isStaticVirtual = (pConstrainedResolvedToken != nullptr && pContextMD != nullptr && pContextMD->IsStatic());
 
     // There is a pathological case where invalid IL refereces __Canon type directly, but there is no dictionary availabled to store the lookup.
```
You can delete this condition. We do not care about bug for bug compatibility for invalid IL in .NET Core.

jkotas replied:
Actually, it does not look right that you had to suppress the check for inlinedLookup. We should never come here with non-shared `pContextMD`. Is this a left-over from earlier versions of the change?

EgorBo replied:
Yeah I think it's a left-over, removed

EgorBo committed:
Address feedback

EgorBo committed:
clean up

EgorBo committed:
Merge branch 'main' into shared-generics-2

EgorBo commented:
/azp run runtime-coreclr outerloop, runtime-coreclr gcstress0x3-gcstress0xc, runtime-nativeaot-outerloop

EgorBo commented:
@jkotas @MichalStrehovsky I think I might need your help with NativeAOT (CoreCLR seems to be fine as is) 😞

DependencyAnalysis is not happy

![image](https://github.com/dotnet/runtime/assets/523221/0fca65a8-cef4-40e6-9635-9a9bc5b0c8f3)

I managed to reduce repro to just `System.Buffers.SharedArrayPool`1<__Canon>.Rent(Int32)` method - it calls ComputeLookup 3 times (from jit's importer)

EgorBo committed:
fix jit assert

EgorBo committed:
Merge branch 'shared-generics-2' of github.com:EgorBo/runtime-1 into shared-generics-2

EgorBo committed:
repro

EgorBo committed:
fix repro

MichalStrehovsky commented:
> @jkotas @MichalStrehovsky I think I might need your help with NativeAOT (CoreCLR seems to be fine as is) 😞

I'll have a look at this in a couple hours but the immediate problem is that we're creating a non-sensical lookup (trying to do a generic lookup with System.Array as the context, that is clearly non-generic).

I submitted https://github.com/dotnet/runtime/pull/99513 that adds an assert that is kicking in while there's still JIT code on the stack for this situation. That might be enough to help you here, but I can have another look.

MichalStrehovsky commented:
I think the issue on native AOT is the same issue that was discussed above (and needs the same fix).

We call `embedGenericHandle`. This computes a lookup within the newly introduced `callerHandle`. In this case, `callerHandle` is a method on a generic type. This makes us note this is `CORINFO_LOOKUP_THISOBJ`.

The JIT then takes that information and feeds it to `getReadyToRunHelper`. But that one didn't get the `callerHandle` treatment so we fall back to `MethodBeingCompiled` as the context once it calls `GetGenericLookupHelper` (and also use the previously computed `CORINFO_LOOKUP_THISOBJ`), but that's wrong and ends up asserting (in the new assert I'm adding) because in this case `MethodBeingCompiled` is a generic method on a non-generic type.

We should be having the same problem for crossgen2 but maybe we just don't have enough testing?

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -12773,9 +12803,20 @@ void Compiler::impInlineInitVars(InlineInfo* pInlineInfo)
             case WellKnownArg::InstParam:
-                // These do not appear in the table of inline arg info; do not include them
+            {
+                InlArgInfo* ctxInfo = getAllocator(CMK_Unknown).allocate<InlArgInfo>(1);
+                memset(ctxInfo, 0, sizeof(*ctxInfo));
```
```suggestion
                InlArgInfo* ctxInfo = new (this, CMK_Inlining) InlArgInfo {};
```

jakobbotsch reviewed (commented):
Did you not need to add handling to `fgInlinePrependStatements`? Doesn't it mean we will end up with a ton of duplicated runtime lookups in some cases? One for each use of the context type parameter in the inlinee.

EgorBo committed:
Add CallerHandle to getReadyToRunHelper

EgorBo commented:
> I think the issue on native AOT is the same issue that was discussed above (and needs the same fix).
> 
> We call `embedGenericHandle`. This computes a lookup within the newly introduced `callerHandle`. In this case, `callerHandle` is a method on a generic type. This makes us note this is `CORINFO_LOOKUP_THISOBJ`.
> 
> The JIT then takes that information and feeds it to `getReadyToRunHelper`. But that one didn't get the `callerHandle` treatment so we fall back to `MethodBeingCompiled` as the context once it calls `GetGenericLookupHelper` (and also use the previously computed `CORINFO_LOOKUP_THISOBJ`), but that's wrong and ends up asserting (in the new assert I'm adding) because in this case `MethodBeingCompiled` is a generic method on a non-generic type.
> 
> We should be having the same problem for crossgen2 but maybe we just don't have enough testing?

Thanks for looking! I've added the callerHandle to `getReadyToRunHelper` and it fixed the assert that you've just added! Unfortunately, it didn't fix the main exception I am hitting with that "Rent" method 😞

EgorBo committed:
add to getReadyToRunDelegateCtorHelper as well

EgorBo committed:
formatting

EgorBo commented:
@jkotas should we for now enable it only for CoreCLR? I presume it will allow us to estimate beneifts (e.g. on dotnet/performance) to see if it's worth the effort on NAOT side

jkotas commented:
> I don't know how to fix that. It looks like this is going to require placing generic dictionary entries that are associated with X into generic dictionary Y. 

This change should not be introducing any new dictionary slots. The dictionary lookups should be chained. The dictionary lookups in the inlinees should go through two or more existing slots.

@MichalStrehovsky Am I missing something?

MichalStrehovsky commented:
> > I don't know how to fix that. It looks like this is going to require placing generic dictionary entries that are associated with X into generic dictionary Y.
> 
> This change should not be introducing any new dictionary slots. The dictionary lookups should be chained. The dictionary lookups in the inlinees should go through two or more existing slots.
> 
> @MichalStrehovsky Am I missing something?

Didn't realize it's chained. If it's chained, it might be fixable. I tried the straightforward fix (add the below into the `if (runtimeDep != null)` of `ShadowConcreteMethod.cs`:

```csharp
                        if (runtimeDep is ReadyToRunGenericLookupFromDictionaryNode lookupFromMethod
                            && (!canonMethod.RequiresInstMethodDescArg() || lookupFromMethod.DictionaryOwner != canonMethod))
                            continue;

                        if (runtimeDep is ReadyToRunGenericLookupFromTypeNode lookupFromType
                            && (!canonMethod.RequiresInstMethodTableArg() || (TypeDesc)lookupFromType.DictionaryOwner != canonMethod.OwningType))
                            continue;
```

It compiles, but we fail at object writing. I suspect this is because there's nothing that would report the dictionary dependencies for the inlined methods if there was no standalone method body generated for it anywhere. We cannot just skip it here, but also we don't have information to map this back to something that can be substituted.

I suspect this is only a problem if we're using the ReadyToRunGenericLookup helpers _and_ we're optimizing. This can only happen if `--noscan -O` was passed to ILC, or multimodule build.

A quick check would be to just drop the `/p:BuildFramework..` from here:

https://github.com/dotnet/runtime/blob/1400c1e7a888ea1e710e5c08d55c800e0b04bf8a/eng/pipelines/runtime.yml#L301

If it's only a problem for that, maybe we can somehow disable this inlining if ReadyToRun helpers are used and we're optimizing. I don't want to lose the ability to `--noscan -O` because it's sometimes useful for troubleshooting

EgorBo commented:
@MichalStrehovsky thansk for investigating! in your snippet what is `canonMethod`? I don't see such variable in `GetStaticDependencies`, is it just `Method` property?

> I suspect this is only a problem if we're using the ReadyToRunGenericLookup helpers and we're optimizing. This can only happen if --noscan -O was passed to ILC, or multimodule build.

Seems to be failing for me as is without additional settings like `-noscan`, here is my RSP:

<details>
  <summary>Click me</summary>

```
C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Text.Json.dll
--jitpath:C:\prj\runtime-main\artifacts\bin\coreclr\windows.x64.Checked\clrjit.dll
--parallelism:32
--verbose
-o:C:\prj\runtime-main\artifacts\tests\coreclr/obj/windows.x64.Checked/Managed/.\build\/NativeAOTFX\System.Text.Json.obj
-r:C:\prj\runtime-main\artifacts\bin\coreclr\windows.x64.Checked\aotsdk\System.Private.CoreLib.dll
-r:C:\prj\runtime-main\artifacts\bin\coreclr\windows.x64.Checked\aotsdk\System.Private.DisabledReflection.dll
-r:C:\prj\runtime-main\artifacts\bin\coreclr\windows.x64.Checked\aotsdk\System.Private.Reflection.Execution.dll
-r:C:\prj\runtime-main\artifacts\bin\coreclr\windows.x64.Checked\aotsdk\System.Private.StackTraceMetadata.dll
-r:C:\prj\runtime-main\artifacts\bin\coreclr\windows.x64.Checked\aotsdk\System.Private.TypeLoader.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\Microsoft.CSharp.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\Microsoft.VisualBasic.Core.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\Microsoft.VisualBasic.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\Microsoft.Win32.Primitives.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\Microsoft.Win32.Registry.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\mscorlib.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\netstandard.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.AppContext.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Buffers.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Collections.Concurrent.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Collections.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Collections.Immutable.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Collections.NonGeneric.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Collections.Specialized.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ComponentModel.Annotations.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ComponentModel.DataAnnotations.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ComponentModel.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ComponentModel.EventBasedAsync.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ComponentModel.Primitives.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ComponentModel.TypeConverter.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Configuration.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Console.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Core.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Data.Common.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Data.DataSetExtensions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Data.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.Contracts.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.Debug.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.DiagnosticSource.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.FileVersionInfo.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.Process.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.StackTrace.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.TextWriterTraceListener.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.Tools.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.TraceSource.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Diagnostics.Tracing.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Drawing.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Drawing.Primitives.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Dynamic.Runtime.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Formats.Asn1.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Formats.Tar.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Globalization.Calendars.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Globalization.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Globalization.Extensions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.Compression.Brotli.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.Compression.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.Compression.FileSystem.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.Compression.ZipFile.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.FileSystem.AccessControl.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.FileSystem.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.FileSystem.DriveInfo.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.FileSystem.Primitives.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.FileSystem.Watcher.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.IsolatedStorage.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.MemoryMappedFiles.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.Pipes.AccessControl.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.Pipes.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.IO.UnmanagedMemoryStream.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Linq.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Linq.Expressions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Linq.Parallel.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Linq.Queryable.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Memory.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Http.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Http.Json.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.HttpListener.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Mail.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.NameResolution.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.NetworkInformation.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Ping.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Primitives.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Quic.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Requests.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Security.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.ServicePoint.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.Sockets.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.WebClient.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.WebHeaderCollection.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.WebProxy.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.WebSockets.Client.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Net.WebSockets.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Numerics.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Numerics.Vectors.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ObjectModel.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Private.DataContractSerialization.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Private.Uri.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Private.Xml.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Private.Xml.Linq.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.DispatchProxy.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.Emit.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.Emit.ILGeneration.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.Emit.Lightweight.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.Extensions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.Metadata.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.Primitives.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Reflection.TypeExtensions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Resources.Reader.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Resources.ResourceManager.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Resources.Writer.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.CompilerServices.Unsafe.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.CompilerServices.VisualC.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Extensions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Handles.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.InteropServices.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.InteropServices.JavaScript.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.InteropServices.RuntimeInformation.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Intrinsics.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Loader.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Numerics.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Serialization.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Serialization.Formatters.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Serialization.Json.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Serialization.Primitives.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Runtime.Serialization.Xml.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.AccessControl.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Claims.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Cryptography.Algorithms.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Cryptography.Cng.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Cryptography.Csp.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Cryptography.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Cryptography.Encoding.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Cryptography.OpenSsl.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Cryptography.Primitives.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Cryptography.X509Certificates.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Principal.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.Principal.Windows.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Security.SecureString.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ServiceModel.Web.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ServiceProcess.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Text.Encoding.CodePages.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Text.Encoding.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Text.Encoding.Extensions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Text.Encodings.Web.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Text.Json.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Text.RegularExpressions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.Channels.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.Overlapped.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.Tasks.Dataflow.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.Tasks.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.Tasks.Extensions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.Tasks.Parallel.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.Thread.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.ThreadPool.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Threading.Timer.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Transactions.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Transactions.Local.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.ValueTuple.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Web.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Web.HttpUtility.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Windows.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.Linq.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.ReaderWriter.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.Serialization.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.XDocument.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.XmlDocument.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.XmlSerializer.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.XPath.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\System.Xml.XPath.XDocument.dll
-r:C:\prj\runtime-main\artifacts\bin\microsoft.netcore.app.runtime.win-x64\Release\runtimes\win-x64\lib\net9.0\WindowsBase.dll
--metadatalog:C:\prj\runtime-main\artifacts\tests\coreclr/obj/windows.x64.Checked/Managed/.\build\/NativeAOTFX\System.Text.Json.metadata.csv
--targetos:win
--targetarch:x64
--multifile
-O
--export-dynamic-symbol:DotNetRuntimeDebugHeader,DATA
--initassembly:System.Private.CoreLib
--initassembly:System.Private.StackTraceMetadata
--initassembly:System.Private.TypeLoader
--initassembly:System.Private.Reflection.Execution
--directpinvoke:System.Globalization.Native
--directpinvoke:System.IO.Compression.Native
--directpinvokelist:C:\prj\runtime-main\artifacts\bin\coreclr\windows.x64.Checked\build\WindowsAPIs.txt
--runtimeknob:System.Linq.Expressions.CanEmitObjectArrayDelegate=false
--runtimeknob:RUNTIME_IDENTIFIER=
--stacktracedata
--scanreflection
--resilient
--generateunmanagedentrypoints:System.Private.CoreLib
--feature:System.Diagnostics.Debugger.IsSupported=false
--feature:System.Threading.ThreadPool.UseWindowsThreadPool=true

```
</details>

EgorBo committed:
Like this (it currently fails)

EgorBo committed:
Update src/coreclr/jit/importer.cpp

Co-authored-by: Jakob Botsch Nielsen <Jakob.botsch.nielsen@gmail.com>

EgorBo committed:
fix build

EgorBo committed:
Fix build

MichalStrehovsky committed:
Revert "Fix build"

This reverts commit 893172c143fb2b0c146d71821c166438a4acec84.

MichalStrehovsky committed:
Revert "fix build"

MichalStrehovsky committed:
Revert "Like this (it currently fails)"

This reverts commit 478ef857655d78b3a99bb861a6522954f06c34e8.

MichalStrehovsky committed:
Revert "fix repro"

This reverts commit 410ce12d0dd576623e71fb3563793645fa94adeb.

MichalStrehovsky committed:
Revert "repro"

This reverts commit 550f8ffa20e20f89edae5b02d7481c5f603bad45.

MichalStrehovsky committed:
Better fix

MichalStrehovsky commented:
I pushed out a better fix to your branch: https://github.com/dotnet/runtime/pull/99265/commits/0f45539a53618ceb13d4169a4fa6674250a568d1

Turns out I just had to sleep on this; the fix is simple: do not inline through generics if ReadyToRun helpers are involved. The new `if` check pretty much only applies to `--multifile` and `-O --noscan`. There's no point in trying to make that one work with inlining.

EgorBo committed:
Address Jakob's feedback

EgorBo commented:
Thanks for fixing the issue @MichalStrehovsky! CI (including outerloop seems to be happy about it).

I presume it's ready for review then? 

PTAL @jakobbotsch cc @dotnet/jit-contrib JIT side, I've addressed your feedback and introduced `fgInsertInlineeArgument` - it's just a copy-pase of the loop body of `fgInlinePrependStatements` for arguments in order to use the same logic for InstParam.

@MichalStrehovsky @jkotas PTAL VM side

Thanks for help everyone! Fingers crossed to see some improvements from this. 
The latest asm-diffs were https://github.com/MihuBot/runtime-utils/issues/278

EgorBo commented:
@MihuBot

jakobbotsch commented on src/coreclr/jit/importer.cpp:
```diff
@@ -12773,9 +12805,19 @@ void Compiler::impInlineInitVars(InlineInfo* pInlineInfo)
             case WellKnownArg::InstParam:
-                // These do not appear in the table of inline arg info; do not include them
+            {
+                InlArgInfo* ctxInfo     = new (this, CMK_Inlining) InlArgInfo{};
+                ctxInfo->arg            = &arg;
+                ctxInfo->argTmpNum      = BAD_VAR_NUM;
+                ctxInfo->argIsLclVar    = arg.GetNode()->OperIs(GT_LCL_VAR);
+                ctxInfo->argIsInvariant = arg.GetNode()->IsInvariant(); // TODO-CQ: consider other trivial cases
```
I.e. it seems safer to reuse the general logic of `impInlineRecordArgInfo`, but with some additional logic to unset some of the flags for known IR shapes.

EgorBo replied:
@jakobbotsch I've pushed a conservative change to always trigger a new temp (except the most popular CNS_INT context). I'd like to see separately (with SPMI) on what I can do to improve the behavior here (and in other TODO-CQ I left)

EgorBo committed:
conservative version

EgorBo committed:
make formatter happy

jakobbotsch reviewed (approved):
JIT side LGTM.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -827,7 +827,7 @@ public void CompileMethod(MethodWithGCInfo methodCodeNodeNeedingCode, Logger log
         }
 
-        private bool getReadyToRunHelper(ref CORINFO_RESOLVED_TOKEN pResolvedToken, ref CORINFO_LOOKUP_KIND pGenericLookupKind, CorInfoHelpFunc id, ref CORINFO_CONST_LOOKUP pLookup)
+        private bool getReadyToRunHelper(ref CORINFO_RESOLVED_TOKEN pResolvedToken, ref CORINFO_LOOKUP_KIND pGenericLookupKind, CorInfoHelpFunc id, CORINFO_METHOD_STRUCT_* callerHandle, ref CORINFO_CONST_LOOKUP pLookup)
```
I'm curious why `callerHandle` can be unused here (we use `pResolvedToken.tokenContext` to get the context instead), but `embedGenericHandle` and `getCallInfo` need the `callerHandle` and don't use context. Since the tests are passing, maybe `callerHandle` is redundant with `tokenContext`?

EgorBo replied:
Switched to callerHandle. From my understanding tokenContext could be a class here in case of inline - that's the whole reason why we propagated CallerHandle.

MichalStrehovsky reviewed (approved):
The ilc and crossgen2 (to the extent that I understand crossgen2) parts lgtm modulo the question.

MichalStrehovsky commented:
> funny enough, it made R2R'd corelib a bit smaller

It's also a small size improvement for native AOT: https://github.com/MichalStrehovsky/rt-sz/issues/2#issuecomment-1990885650

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -2894,6 +2895,7 @@ void CEEInfo::embedGenericHandle(
                                                   pResolvedToken,
                                                   NULL,
                                                   pTemplateMD,
+                                                  (MethodDesc*)callerHandle,
```
```suggestion
                                                  GetMethod(callerHandle),
```
This is the style used in this file instead of C-casts. (Fix all places.)

jkotas commented:
`METHOD_BEING_COMPILED_CONTEXT` should not be needed anymore and it can be deleted. Did you run into problems with deleting it?

EgorBo commented:
> `METHOD_BEING_COMPILED_CONTEXT` should not be needed anymore and it can be deleted. Did you run into problems with deleting it?

I found two places I don't have a good understanding what they're needed for:
1) https://github.com/dotnet/runtime/blob/d36beb7b0a3936d18062ca89572d49590fdd445a/src/coreclr/jit/inline.cpp#L1267 - it sees to be used to detect recursive inlining or something like that
2) https://github.com/dotnet/runtime/blob/d36beb7b0a3936d18062ca89572d49590fdd445a/src/coreclr/vm/jitinterface.cpp#L5303

I might delete the 1st one but I think I'll need an SPMI collection for that so ideally after this PR is merged if you don't mind

EgorBo committed:
use GetMethod

EgorBo committed:
Merge branch 'main' of github.com:dotnet/runtime into shared-generics-2

# Conflicts:
#	src/coreclr/inc/jiteeversionguid.h
#	src/coreclr/tools/Common/JitInterface/CorInfoImpl_generated.cs

EgorBo committed:
Address feedback

jkotas commented:
> I might delete the 1st one but I think I'll need an SPMI collection for that so ideally after this PR is merged if you don't mind

It is fine with me to do it in a follow up PR. Note that deleting `METHOD_BEING_COMPILED_CONTEXT` is JIT/EE interface breaking change, so you need to rev the GUID.

The second place is what enabled the limited shared generic inlining optimization that should be replaced by your change. It checks for the happy path that the optimization was capable of handling.

EgorBo commented:
/azp run runtime-coreclr crossgen2

EgorBo commented:
> The second place is what enabled the limited shared generic inlining that should be superseded by your change. It checks for the happy path that the optimization was capable of handling.

@jkotas do you mean it should be just `pResult->contextHandle = callerHandle` without those conditions, rights?

jkotas commented:
It should be `pResult->contextHandle = MAKE_CLASSCONTEXT(exactType.AsPtr());` that is there a few lines before the if check. The if check and `pResult->contextHandle = METHOD_BEING_COMPILED_CONTEXT()` should be deleted without replacement.

> pResult->contextHandle = callerHandleAsContext without those conditions, rights?

This would not work well - it would be a code quality regression. The context needs to be `exactType` to carry as much instantiation details as possible into the inlinee.

EgorBo committed:
Address Jan's feedback

EgorBo commented:
@jkotas anything else (assuming CI passes)?
I already have a couple of ideas for a follow up to clean up/improve CQ, so going to include `METHOD_BEING_COMPILED_CONTEXT` there

jkotas reviewed (approved):
Looks great! Thank you

EgorBo commented:
Thanks for help!

EgorBo commented:
Found at least 80 benchmarks improved by 10% or more: https://gist.github.com/EgorBo/b6424f7118ff176682f63875d89fb52e

EgorBo commented:
Improvements on x64:

- https://github.com/dotnet/perf-autofiling-issues/issues/31364
- https://github.com/dotnet/perf-autofiling-issues/issues/31398
- https://github.com/dotnet/perf-autofiling-issues/issues/31429

Regressions on x64:

- https://github.com/dotnet/perf-autofiling-issues/issues/31420
